test/integration/http_integration.cc

```diff
@@ -1274,7 +1279,104 @@ void HttpIntegrationTest::testUpstreamDisconnectWithTwoRequests() {
   test_server_->waitForCounterGe("cluster.cluster_0.upstream_rq_200", 2);
 }
 
-void HttpIntegrationTest::testTwoRequests() {
+// This filter exists to synthetically test network backup by faking TCP
+// conneciton back-up when an encode is finished.
+// Allows regression tests for https://github.com/envoyproxy/envoy/issues/4541
+// TODO(alyssawilk) move this somewhere general and turn it up for more tests.
+class TestPauseFilter : public PassThroughFilter {
+public:
+  // A wrapper around a TimerPtr which can be passed to the timer on creation,
+  // so it can delete itself when called.
+  struct SelfOwningTimer {
+    Event::TimerPtr timer_;
+  };
+
+  // Pass in a global "is this the first stream to finish encoding" to ensure
+  // the Network::Connection is "blocked" exactly once.
+  TestPauseFilter(absl::Mutex& encode_lock, bool& first_encode_ref)
+      : encode_lock_(encode_lock), first_encode_ref_(first_encode_ref) {}
+
+  Http::FilterDataStatus encodeData(Buffer::Instance& buf, bool end_stream) override {
+    absl::WriterMutexLock m(&encode_lock_);
+    // If this is the first time any stream has finished encoding, force the
+    // connection to a high watermark state, and set an alarm to force low
+    // watermark state.
+    if (end_stream && first_encode_ref_) {
+      forceHighWatermark();
+    }
+    return PassThroughFilter::encodeData(buf, end_stream);
+  }
+
+  void forceHighWatermark() {
+    first_encode_ref_ = false; // Make sure we only do this once.
+    Network::ConnectionImpl* connection_impl = connection();
+    connection_impl->onHighWatermark(); // Force high watermark state.
+
+    // Unfortunately this filter is about to self destruct, so the timer
+    // to lower the watermark state must outlive the filter.
+    //
+    // Create a "self-owning" timer and pass it an absolute pointer to the
+    // connection. The timer will delete itself on destruction.
+    //
+    // It is up to the users of this filter to make sure the connection_impl
+    // will outlive the timer.
+    SelfOwningTimer* timer_container = new SelfOwningTimer;
+    timer_container->timer_ =
+        decoder_callbacks_->dispatcher().createTimer([connection_impl, timer_container]() -> void {
+          ENVOY_LOG_MISC(debug, "Forcing watermark low");
+          connection_impl->onLowWatermark();
+          delete timer_container;
+        });
+    timer_container->timer_->enableTimer(std::chrono::milliseconds(50));
```

Comment:
This seems surely destined to cause flakes. Any thoughts on how to make this less flaky? Is there anything we can do with @jmarantz time source stuff here?

Reply:
FWIW this should work fine when HttpIntegrationTest is constructed with a simTime().

If this path is exercised when HttpIntegrationTest is constructed with realTIme() I would expect the behavior to vary.

Reply:
Howso?    The whole point is to semi-randomly block and semi-randomly unblock the socket.  AFIK it'll only cause flakes if we actually block the network socket which again I was completely unable to do on all of our linux builds, with large responses and artificially small buffers.

Reply:
It might be fine, but just from my quick look it seems like it could flake if the 2nd stream comes in either before or after we unblock? Or is the test not meant to be deterministic?

Reply:
Yeah, it would not guarantee a repro if the second stream arrived unexpectedly.  In practice it happened to fail all the time but I agree on a slow system the repro would not be deterministic.

If we prefer deterministic testing I can fix that by tracking the number of decodes, and unblocking after the second stream is created.  I kind of like fuzz testing timing but I wouldn't be super sad to see the self-owning timer and lifetime issues go away

Reply:
IMO having a deterministic test would be better if we had one test, but perhaps if it's possible to do determinism we can also leave this test as is? Will defer to you on how to handle.

Reply:
Actually I think I like this - I can do "random" encode/decode logic when I generalize for other tess.